torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -328,117 +295,102 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+                    if local_param_index in s["state"].keys():
+                        unordered_state[local_index_to_param_id[local_param_index]] = s["state"][local_param_index]
+
+        # - save the states in the expected order, meaning the one from the param_groups
+        for i, param in enumerate(chain(*(g["params"] for g in self.param_groups))):
+            if id(param) in unordered_state.keys():
+                state_dict["state"][i] = unordered_state[id(param)]
+
+        return state_dict
 
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
         Arguments:
-            rank (int): rank to get local_state_dict for
-            state_dict (dict): global state_dict
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
         """
-        param_groups = state_dict["param_groups"][state_dict["partition"][rank][0] : state_dict["partition"][rank][1]]
-        return {"state": state_dict["state"][rank], "param_groups": param_groups}
+
+        # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
+        # we work around that here by using the fact that the params are ordered as in the param_groups
+
+        # Param index to param map
+        global_id_map = {i: p for i, p in enumerate(chain(*(g["params"] for g in self.param_groups)))}
+
+        # Prune the state_dict from the states which this rank does not own, then normal base load
+        other_state = []
+        for i_param, key in enumerate(state_dict["state"].keys()):
+            # Check that this rank owns this param, if not remove from the state
+            param = global_id_map[i_param]
+            if self.param_to_rank[param] != self.rank:
+                other_state.append(key)
+
+        # Keep the state in place in order not to break the following enumerations, but wipe the contents
+        for other_parameter in other_state:
+            state_dict["state"][other_parameter] = None
+
+        super().load_state_dict(state_dict)
+
+        # Set the sharded optimizer state.
+        # Keep the original type (not respected by PyTorch which casts to the model type)
+        for k, (_, v) in enumerate(state_dict["state"].items()):
+            if k in global_id_map:
+                param = global_id_map[k]
+
+                # Only add this state to the sharded optimizer if it owns this param
+                for pg in self.optim.param_groups:
+                    if id(param) in map(lambda x: id(x), pg["params"]):
+                        self.optim.state[param] = _recursive_copy_to_device(v, non_blocking=True, device=param.device)
+
+        # Update the param_group keys
+        for fpg, pg in zip(state_dict["param_groups"], self.param_groups):
+            for key in fpg.keys():
+                if key != "params":
+                    pg[key] = fpg[key]
```

Comment:
it does apply to all the params exposed in the param_groups, because the model is not sharded here, and frameworks (at least Classy) assume that everything here is visible. I'll add a note, you're right